src/convolution.jl

```diff
@@ -0,0 +1,100 @@
+"""
+    convolve(d1::T, d2::T) where T<:Distribution -> Distribution
+
+Convolve two distributions of the same type to yield the distribution corresponding to the
+sum of independent random variables drawn from the underlying distributions.
+
+The function is only defined in the cases where the convolution has a closed form as
+defined here https://en.wikipedia.org/wiki/List_of_convolutions_of_probability_distributions
+
+* Bernoulli
+* Binomial
+* Negative Binomial
+* Geometric
+* Poisson
+* Normal
+* Cauchy
+* Chisq
+* Exponential
+* Gamma
+* Multi-variate Normal
+"""
+#convolve(d1, d2) = throw(ArgumentError("`convolve` is not defined for $(typeof(d1)) and $(typeof(d2))"))
+
+# discrete univariate
+function convolve(d1::Bernoulli, d2::Bernoulli)
+    _check_convolution_args(d1.p, d2.p)
+    return Binomial(2, d1.p)
+end
+
+function convolve(d1::Binomial, d2::Binomial)
+    _check_convolution_args(d1.p, d2.p)
+    return Binomial(d1.n + d2.n, d1.p)
+end
+
+function convolve(d1::NegativeBinomial, d2::NegativeBinomial)
+    _check_convolution_args(d1.p, d2.p)
+    return NegativeBinomial(d1.r + d2.r, d1.p)
+end
+
+function convolve(d1::Geometric, d2::Geometric)
+    _check_convolution_args(d1.p, d2.p)
+    return NegativeBinomial(2, d1.p)
+end
+
+convolve(d1::Poisson, d2::Poisson) =  Poisson(d1.λ + d2.λ)
+
+
+# continuous univariate
+convolve(d1::Normal, d2::Normal) = Normal(d1.μ + d2.μ, hypot(d1.σ, d2.σ))
+convolve(d1::Cauchy{T}, d2::Cauchy{T}) where T = Cauchy(d1.μ + d2.μ, d1.σ + d2.σ)
+convolve(d1::Chisq, d2::Chisq) = Chisq(d1.ν + d2.ν)
+
+function convolve(d1::Exponential, d2::Exponential)
+    _check_convolution_args(d1.θ, d2.θ)
+    return Gamma(2, d1.θ)
+end
+
+function convolve(d1::Gamma, d2::Gamma)
+    _check_convolution_args(d1.θ, d2.θ)
+    return Gamma(d1.α + d2.α, d1.θ)
+end
+
+# continuous multivariate
+function convolve(
+    d1::Union{IsoNormal, ZeroMeanIsoNormal},
+    d2::Union{IsoNormal, ZeroMeanIsoNormal},
+    )
+    _check_convolution_shape(d1, d2)
+    return MvNormal(d1.μ .+ d2.μ, d1.Σ.value + d2.Σ.value)
+end
+
+function convolve(
+    d1::Union{DiagNormal, ZeroMeanDiagNormal},
+    d2::Union{DiagNormal, ZeroMeanDiagNormal},
+    )
+    _check_convolution_shape(d1, d2)
+    return MvNormal(d1.μ .+ d2.μ, d1.Σ.diag + d2.Σ.diag)
+end
+
+function convolve(
+    d1::Union{FullNormal, ZeroMeanFullNormal},
+    d2::Union{FullNormal, ZeroMeanFullNormal},
+    )
+    _check_convolution_shape(d1, d2)
+    return MvNormal(d1.μ .+ d2.μ, d1.Σ.mat + d2.Σ.mat)
+end
+
+function convolve(d1::MvNormal, d2::MvNormal)
```

Comment:
I wonder if this function is sufficient for all MvNormal cases or are the above functions also necessary? Are there any expected performance benefits from dispatching on the sub-type?

Reply:
I suspect calling `Matrix()` on some very large diagonal matrices could be an unnecessary bottleneck?

Reply:
Yes, I think the specialised methods you added have performance benefits, because they can avoid the `Matrix` call

but also I think they reveal that the `MvNormal` types are a bit wierd (https://github.com/JuliaStats/Distributions.jl/issues/921)

Reply:
I see the `Matrix` here is for handling the diagonal + matrix case 👍 

I'll note that the code here does not handle `Int`s: this is the only method that anything with `Int`s will hit (becuase #921)... without the `Matrix` call we'd get an `Inexact` error
```julia
julia> z = MvNormal([1, 1])
MvNormal{Int64,PDiagMat{Int64,Array{Int64,1}},Distributions.ZeroVector{Int64}}(
dim: 2
μ: [0, 0]
Σ: [1 0; 0 1]
)

julia> MvNormal(z.μ .+ z.μ, z.Σ + z.Σ)
ERROR: InexactError: Int64(0.5)
...
```

The `Matrix` call saves us from that... But we hit a problem in PDMats (I guess i'll open an issue there -- but another vote for ditching PDMats https://github.com/JuliaStats/Distributions.jl/issues/688#issuecomment-346145780)

```julia
julia> MvNormal(z.μ .+ z.μ, Matrix(z.Σ) + Matrix(z.Σ))
ERROR: MethodError: no method matching PDMat{Int64,Array{Int64,2}}(::Int64, ::Array{Int64,2}, ::Cholesky{Float64,Array{Float64,2}})
Closest candidates are:
  PDMat{Int64,Array{Int64,2}}(::Int64, ::AbstractArray{T,2}, ::Cholesky{T,S}) where {T, S} at /Users/nick/.julia/packages/PDMats/AObTs/src/pdmat.jl:9
Stacktrace:
 [1] PDMat(::Array{Int64,2}, ::Cholesky{Float64,Array{Float64,2}}) at /Users/nick/.julia/packages/PDMats/AObTs/src/pdmat.jl:16
 [2] PDMat(::Array{Int64,2}) at /Users/nick/.julia/packages/PDMats/AObTs/src/pdmat.jl:19
 [3] MvNormal(::Array{Int64,1}, ::Array{Int64,2}) at /Users/nick/.julia/packages/Distributions/Iltex/src/multivariate/mvnormal.jl:203
 [4] top-level scope at none:0
```

Reply:
That is weird to restrict to just `Float64`...

Maybe to avoid the inexact error I could do 

```julia
function convolve(
    d1::Union{IsoNormal{T}, ZeroMeanIsoNormal{T}, DiagNormal{T}, ZeroMeanDiagNormal{T}},
    d2::Union{IsoNormal{T}, ZeroMeanIsoNormal{T}, DiagNormal{T}, ZeroMeanDiagNormal{T}},
    ) where T<:Float64
```

So that anything with `Int` will get sent to the other method?

Reply:
It's very weird that they would just be defined over `Float64`...
It would be great if https://github.com/JuliaStats/Distributions.jl/issues/921 got resolved.